File: 0078-subsets/0078-subsets.py
class Solution(object):
    def subsets(self, nums):
        res = []
        subset = []
        
        def dfs(i):
            if i >= len(nums):
                res.append(subset.copy())
                return  # Add this return statement to exit the function
            
            subset.append(nums[i])  # Include the current element in the subset
            dfs(i + 1)  # Recur to include the next element
            
            subset.pop()  # Backtrack and remove the current element
            dfs(i + 1)  # Recur without including the current element
        
        dfs(0)
        return res

File: 0078-subsets/0078-subsets.cpp
using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int>curSet;
        vector<vector<int>>subSets;
        helper(nums, 0, curSet, subSets);
        return subSets;
    }

    void helper(vector<int>& nums, int i, vector<int>& curSet, vector<vector<int>>& subSets) {
        if (i >= nums.size()) {
            subSets.push_back(vector<int>(curSet));
            return;
        }
        
        curSet.push_back(nums[i]);
        helper(nums, i + 1, curSet, subSets);
        curSet.pop_back();



        helper(nums, i + 1, curSet, subSets);
    }
};
